File: jvm/leetcode_scala/src/main/scala/leetcode/editor/cn/[872]叶子相似的树.scala
package leetcode.editor.cn
object LeafSimilarTrees {
  import scala.util.control.TailCalls.*
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for a binary tree node. class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) { var value: Int = _value var left: TreeNode = _left var right: TreeNode =
    * _right }
    */
  object Solution {
    import scala.util.boundary
    def leafSimilar(root1: TreeNode, root2: TreeNode): Boolean = {
      val itr1 = LeafIterator(root1)
      val itr2 = LeafIterator(root2)

      boundary:
        while itr1.hasNext && itr2.hasNext do if itr1.next.value != itr2.next.value then boundary.break(false)

        itr1.hasNext == itr2.hasNext
    }

    import scala.collection.mutable
    class LeafIterator(node: TreeNode) extends Iterator[TreeNode] {
      private val stack = mutable.Stack[TreeNode](node)

      override def hasNext: Boolean = stack.nonEmpty

      override def next(): TreeNode = {
        boundary:
          while stack.nonEmpty  do
            val cur = stack.pop()
            if cur.left == null && cur.right == null then boundary.break(cur)
            else
              if cur.right != null then stack.push(cur.right)
              if cur.left != null then stack.push(cur.left)
          null
      }
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

  import scala.util.control.TailCalls.*

  /** 栈安全的State Monad。可用于leetcode的用fp方式来解决问题。
    */
  case class State[S, A](run: S => TailRec[(S, A)]) {
    def evalState(s: S): A = run(s).result._2
    def execState(s: S): S = run(s).result._1
    def map[B](ab: A => B): State[S, B] =
      flatMap(a => State(s => done((s, ab(a)))))

    def flatMap[B](afb: A => State[S, B]): State[S, B] =
      State(s =>
        tailcall(
          run(s).flatMap { case (s, a) => tailcall(afb(a).run(s)) }
        )
      )
  }
  object StateOps {
    def point[S, A](a: A): State[S, A] = State(s => done((s, a)))

    def get[S]: State[S, S] = State(s => done((s, s)))

    def put[S](s: S): State[S, Unit] = State(_ => done((s, ())))

    def modify[S](ss: S => S): State[S, Unit] =
      for { s <- get[S]; _ <- put(ss(s)) } yield ()
  }
}

File: cpp/leetcode/editor/cn/[872]叶子相似的树.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <functional>
#include <queue>

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode *root1, TreeNode *root2) {
        LeafContainer cnt1(root1);
        LeafContainer cnt2(root2);

        return std::equal(cnt1.begin(), cnt1.end(), cnt2.begin(), cnt2.end(), [](auto left, auto right) {
            if (left == nullptr || right == nullptr) return false;
            return left->val == right->val;
        });
    }

private:
    class LeafContainer {
    public:
        class LeafIterator {
        public:
            using iterator_category = std::forward_iterator_tag;
            using difference_type = std::ptrdiff_t;
            using value_type = TreeNode *;
            using pointer = value_type *;
            using reference = value_type &;


            LeafIterator(LeafContainer *container, bool end) : container{container}, current{nullptr} {
                if (container != nullptr && !end)
                    this->operator++();
            }

            bool operator==(const LeafIterator &other) const {
                return container == other.container && current == other.current;
            }

            reference operator*() {
                return current;
            }

            LeafIterator &operator++() {
                if (container == nullptr) return *this;
                while (!container->stack.empty()) {
                    auto cur = container->stack.top();
                    container->stack.pop();
                    if (cur->left == nullptr && cur->right == nullptr) {
                        current = cur;
                        return *this;
                    } else {
                        if (cur->right != nullptr) container->stack.push(cur->right);
                        if (cur->left != nullptr) container->stack.push(cur->left);
                    }
                }
                current = nullptr;
                return *this;
            }


        private:
            value_type current;
            LeafContainer *container;
        };

        LeafContainer(TreeNode *root) {
            if (root != nullptr)
                stack.push(root);
        }

        LeafIterator begin() {
            return LeafIterator(this, false);
        }

        LeafIterator end() {
            return LeafIterator(this, true);
        }

    private:
        std::stack<TreeNode *> stack;
    };
};
//leetcode submit region end(Prohibit modification and deletion)

#include <iostream>

int main() {
    auto root1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    auto root2 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    Solution s;
    std::cout << s.leafSimilar(root1, root2);
}